### PR TITLE
Changes required for running the hiway-chef cookbooks

### PR DESCRIPTION
**Pull request by marcbux**
Aside from a couple insignificant formatting and parameter changes, the changes proposed in this pull request are as follows:

(1) I found the Apache URL for downloading Hadoop to not work reliably. Every 20th or so time I ran kitchen, the download would commence but the checksum would not match, thus crashing the VM setup. I therefore switched to a different URL. The problem then was that the regular Hadoop mirrors only provide Hadoop back until version 2.5.1. Therefore, I’ve put in a version check that chooses a mirror based on version number (i.e., if version < 2.5.1, use Apache archives mirror, otherwise use FU Berlin mirror). This change is obviously not required by Hi-WAY whatsoever, but for me it has increased stability of the cookbooks for Hadoop version >= 2.5.1.

(2) I’ve put in place the CapacityScheduler with policy DominantResourceScheduler as the default scheduler for Hadoop. This enables CPU management in YARN (see http://blog.cloudera.com/blog/2013/12/managing-multiple-resources-in-hadoop-2-with-yarn/), which Hi-WAY makes use of. If you wish to use a scheduler w/o CPU management (e.g., FifoScheduler), and still use the hiway-chef recipes, you will have to set [:hiway][:hiway][:am][:vcores] and [:hiway][:hiway][:worker][:vcores] to 1 or Hi-WAY will not get the containers it requests and thus not work.

(3) I’ve made several parameters visible to Karamel via the metadata.rb. In particular, these are [:hadoop][:yarn][:max_vcores], [:hadoop][:num_replicas], [:hadoop][:container_cleanup_delay_sec]. hiway-chef makes use of these (at least the first two) in my yaml files, so they have become a necessity.

### CODE REVIEW

File: templates/default/capacity-scheduler.xml.erb [hunk 1]
```diff
@@ -33,7 +33,7 @@
 
   <property>
     <name>yarn.scheduler.capacity.resource-calculator</name>
-    <value>org.apache.hadoop.yarn.util.resource.DefaultResourceCalculator</value>
+    <value>org.apache.hadoop.yarn.util.resource.DominantResourceCalculator</value>
```
Comment (jimdowling) [inline templates/default/capacity-scheduler.xml.erb]: The DominantResourceCalculator is not available in hadoop 2.4, http://hadoop.apache.org/docs/r2.4.1/hadoop-yarn/hadoop-yarn-site/CapacityScheduler.html, so you can override this in your yml files. Can you make this a parameter, and keep the defaultResourceCalculator?

File: attributes/default.rb [hunk 2]
```diff
@@ -11,13 +11,18 @@
...
+default[:hadoop][:bin_dir]                 = "#{node[:hadoop][:home]}/bin"
 default[:hadoop][:data_dir]                = "/var/data/hadoop"
 default[:hadoop][:dn][:data_dir]           = "#{node[:hadoop][:data_dir]}/hdfs/dn"
 default[:hadoop][:nn][:name_dir]           = "#{node[:hadoop][:data_dir]}/hdfs/nn"
 
-default[:hadoop][:download_url]            = "https://archive.apache.org/dist/hadoop/core/hadoop-#{node[:hadoop][:version]}/hadoop-#{node[:hadoop][:version]}.tar.gz"
-#default[:hadoop][:download_url]            = "#{node[:download_url]}/hadoop-#{node[:hadoop][:version]}.tar.gz"
+default[:hadoop][:download_url]            = "ftp://ftp.fu-berlin.de/unix/www/apache/hadoop/common/hadoop-#{node[:hadoop][:version]}/hadoop-#{node[:hadoop][:version]}.tar.gz"
+version = "#{node[:hadoop][:version]}"
+version.gsub!('.', '')
```
Comment (jimdowling) [inline attributes/default.rb]: This is taking recent versions for fu-berlin, and old versions from hadoop archive. Is that correct?

File: templates/default/capacity-scheduler.xml.erb (hunk 1 above)
Comment (marcbux) [inline templates/default/capacity-scheduler.xml.erb]: Good point. I'll do as you suggested, but I'll also add a check that will
set hiway's vcores parameters to 1 if the DominantResourceCalculator is not
used.

I'll create a new pull request as soon as I'm done with it.

Comment (jimdowling) [inline templates/default/capacity-scheduler.xml.erb]: Great.

On 2015-11-10 15:43, marcbux wrote:
>
> In templates/default/capacity-scheduler.xml.erb 
> <https://github.com/hopshadoop/apache-hadoop-chef/pull/6#discussion_r44413332>:
>
> > @@ -33,7 +33,7 @@
> >
> >    <property>
> >      <name>yarn.scheduler.capacity.resource-calculator</name>
> > -    <value>org.apache.hadoop.yarn.util.resource.DefaultResourceCalculator</value>
> > +    <value>org.apache.hadoop.yarn.util.resource.DominantResourceCalculator</value>
> Good point. I'll do as you suggested, but I'll also add a check that 
> will set hiway's vcores parameters to 1 if the 
> DominantResourceCalculator is not used. I'll create a new pull request 
> as soon as I'm done with it.
> …
> On Tue, Nov 10, 2015 at 12:51 PM, Jim Dowling 
> <notifications@github.com> wrote: In 
> templates/default/capacity-scheduler.xml.erb <#6 (comment) 
> <https://github.com/hopshadoop/apache-hadoop-chef/pull/6#discussion_r44397559>> 
> : > @@ -33,7 +33,7 @@ > > <property> > 
> <name>yarn.scheduler.capacity.resource-calculator</name> > - 
> <value>org.apache.hadoop.yarn.util.resource.DefaultResourceCalculator</value> 
> > + 
> <value>org.apache.hadoop.yarn.util.resource.DominantResourceCalculator</value> 
> The DominantResourceCalculator is not available in hadoop 2.4, 
> http://hadoop.apache.org/docs/r2.4.1/hadoop-yarn/hadoop-yarn-site/CapacityScheduler.html, 
> so you can override this in your yml files. Can you make this a 
> parameter, and keep the defaultResourceCalculator? — Reply to this 
> email directly or view it on GitHub 
> <https://github.com/hopshadoop/apache-hadoop-chef/pull/6/files#r44397559>. 
>
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/hopshadoop/apache-hadoop-chef/pull/6/files#r44413332>.
>

-- 
-----
Dr. Jim Dowling
Senior Researcher, SICS - Swedish ICT
Associate Prof, KTH - Royal Institute of Technology

File: attributes/default.rb (hunk 2 above)
Comment (marcbux) [inline attributes/default.rb]: That's correct. Actually though, it's kind of a hack. Let me change the
behavior such that there is a primary and a secondary URL and the primary
is always checked first. If the binaries can't be obtained from the primary
URL, check the secondary instead. I think that makes more sense. Do you
agree?

I'll create a new pull request as soon as I'm done with it.

Comment (jimdowling) [inline attributes/default.rb]: Ok

-----Original Message-----

**Comment by marcbux** [comment]
Ok, the discussed changes are in.

**Pull request merged**